### ISSUE TITLE
Problem with MCTS

### ISSUE THREAD

**Issue by sethtroisi** [tensorflow/minigo#379]
the PUCT formula (From AlphaGo Zero pg 26) is `cpuct * policy[move] * sqrt(sum(child_N)) / (1 +child_N[move])`

Our current implementation ([mcts.py#L106](https://github.com/tensorflow/minigo/blob/master/mcts.py#L106)) uses `sqrt(1 + self.N)`

Problems:
1. `+ 1` not mentioned in AlphaGo
1. `self.N != sum(child_N)`, this cut shortcut is generally off by 1 (as the initial expansion of the root increments self.N but not any of child_N)
1. Minigo [increments edge count](https://github.com/tensorflow/minigo/blob/master/mcts.py#L138) prior to select instead of during backup, so that `self.N = sum(child_N) + 2` when action_scores are calculated.

All of this results in `sqrt(sum(child_N) + 3)` instead of `sqrt(sum(child_N))`, which is clearly wrong.
To muddy this conversation, `sqrt(sum(child_N) +1)` seems to make more sense for because otherwise when there's no visits `sqrt(sum(child_N))` equals 0 and argmax will need a tiebreaker.

**Comment by sethtroisi** [comment]
```
minigo-master v minigo-mcts-fix (48/500 games)
board size: 19   komi: 7.5
                  wins              black         white       avg cpu
minigo-master       20 41.67%       10 41.67%     10 41.67%     57.21
minigo-mcts-fix     28 58.33%       14 58.33%     14 58.33%     57.14
                                    24 50.00%     24 50.00%
```

From

```
competition_type = 'playoff'

description = """ Testing various minigo models """

record_games = True
stderr_to_log = True

BASE_FLAGS = " --resign_threshold=-0.88 --num_readouts=400 "
CC_FLAGS = " --mode=gtp --virtual_losses=8 --soft_pick=false --inject_noise=false --                 courtesy_pass=false "
ALL_FLAGS = CC_FLAGS + BASE_FLAGS

def cc_player(folder):
    return Player(
        "bazel-bin/cc/main" + ALL_FLAGS + " --model=models/000303-olympus.pb",
        cwd="~/Projects/" + folder + "/")

players = {
    'minigo-master'   : cc_player('minigo'),
    'minigo-mcts-fix' : cc_player('minigo2'),
}

board_size = 19
komi = 7.5

matchups = [
    Matchup('minigo-master', 'minigo-mcts-fix', alternating=True, scorer='players',                  number_of_games=500),
]
```

**Comment by amj** [comment]
i'm really glad to see these results and they look promising, but i'm not sure this is sufficient evidence.  as the fix changes the interaction of c_puct, you'd need to re-tune c_puct for the new mcts code, right?

(also, why 303-olympus? 20x128?)

**Comment by sethtroisi** [comment]
303 because it already downloaded. I'm doing a 2nd test with 451 (also v7) right now.

* I included the results to make sure nothing it wasn't a big regression.
* This is more correct (in terms of reproducing the paper), and we don't seem to make changes based on what has the highest winrate (See First Play Urgency).
* It's unclear how this would interact with cpuct and RL; And it more clearly matches the AG/AGZ paper so I think we should default to it (otherwise I'm going back to argueing for FPU)

**Comment by amj** [comment]
I agree this seems more correct! :)  i think we should get it in for v10.  Should we re-tune c_puct?

**Comment by brilee** [comment]
Also agreed that retuning c_puct seems advisable.

**Comment by sethtroisi** [comment]
With 800 readouts
```
minigo-master v minigo-mcts-fix (111/500 games)
board size: 19   komi: 7.5
                  wins              black         white       avg cpu
minigo-master       55 49.55%       33 58.93%     22 40.00%    453.63
minigo-mcts-fix     56 50.45%       33 60.00%     23 41.07%    453.88
                                    66 59.46%     45 40.54%
```

**Comment by amj** [comment]
fixed by #392

**Comment by sethtroisi** [comment]
Andrew's patch superseded my patches:
https://github.com/tensorflow/minigo/pull/382
and
https://github.com/tensorflow/minigo/pull/386
